swarm/shed/index.go

```diff
@@ -0,0 +1,264 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package shed
+
+import (
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// IndexItem holds fields relevant to Swarm Chunk data and metadata.
+// All information required for swarm storage and operations
+// on that storage must be defined here.
+// This structure is logically connected to swarm storage,
+// the only part of this package that is not generalized,
+// mostly for performance reasons.
+//
+// IndexItem is a type that is used for retrieving, storing and encoding
+// chunk data and metadata. It is passed as an argument to Index encoding
+// functions, get function and put function.
+// But it is also returned with additional data from get function call
+// and as the argument in iterator function definition.
+type IndexItem struct {
+	Address         []byte
+	Data            []byte
+	AccessTimestamp int64
+	StoreTimestamp  int64
+	// UseMockStore is a pointer to identify
+	// an unset state of the field in Join function.
```

Comment:
w00t??

Reply:
@justelad for `Join` (or `merge`) to work you need nullable types in this struct

Reply:
I forgot to remove this field as it is not needed.

As the bool have only two states, we need another one to identify if it is set or not. In some case a default value can be used for this, like for []byte and int where nil and 0 can represent unset state, but for bool, we do not know if false is unset or set state, so a pointer to bool is providing a third state nil for that purpose.